src/main/java/com/englishtown/bitbucket/hook/MirrorBucketProcessor.java

```diff
@@ -0,0 +1,147 @@
+        // settings (e.g. username/password) have been changed since the first request was
+        // queued, we process the _last_ request in the list. Since mirroring pushes all of
+        // the configured refspecs, any single request should roll up changes from any number
+        // of requests
+        MirrorRequest request = requests.get(requests.size() - 1);
+
+        securityService.withPermission(Permission.REPO_READ, "Mirror changes")
+                .call(() -> {
+                    Repository repository = repositoryService.getById(request.getRepositoryId());
+                    if (repository == null) {
+                        log.debug("{}: Repository has been deleted", request.getRepositoryId());
+                        return null;
+                    }
+                    if (repositoryService.isEmpty(repository)) {
+                        log.debug("{}: The repository is empty", repository);
+                        return null;
+                    }
+                    runMirrorCommand(request.getSettings(), repository);
+
+                    return null;
+                });
+    }
+
+    private void runMirrorCommand(MirrorSettings settings, Repository repository) {
+        log.warn("{}: Preparing to push changes to mirror", repository);
+
+        String password = passwordEncryptor.decrypt(settings.password);
+        String authenticatedUrl = getAuthenticatedUrl(settings.mirrorRepoUrl, settings.username, password);
+
+        // Call push command with the prune flag and refspecs for heads and tags
+        // Do not use the mirror flag as pull-request refs are included
+        ScmCommandBuilder<?> builder = scmService.createBuilder(repository)
+                .command("push")
+                .argument("--prune") // this deletes locally deleted branches
+                .argument(authenticatedUrl)
+                .argument("--force");
+
+        // Use an atomic transaction to have a consistent state
+        if (settings.atomic) {
+            builder.argument("--atomic");
+        }
+
+        // Add refspec args
+        String refspecs = Strings.isNullOrEmpty(settings.refspec) ? DEFAULT_REFSPEC : settings.refspec;
+        for (String refspec : refspecs.split("\\s|\\n")) {
+            if (!Strings.isNullOrEmpty(refspec)) {
+                builder.argument(refspec);
+            }
+        }
+
+        // Add tags refspec
+        if (settings.tags) {
+            builder.argument("+refs/tags/*:refs/tags/*");
+        }
+        // Add notes refspec
+        if (settings.notes) {
+            builder.argument("+refs/notes/*:refs/notes/*");
+        }
+
+        PasswordHandler passwordHandler = new PasswordHandler(settings.password,
+                new GitCommandExitHandler(i18nService, repository));
+
+        Command<String> command = builder.errorHandler(passwordHandler)
+                .exitHandler(passwordHandler)
+                .build(passwordHandler);
+        command.setTimeout(timeout);
```

Comment:
This fixes #39 by allowing administrators to configure a timeout in `bitbucket.properties`. The default timeout, 2 minutes, matches what Bitbucket Server is currently applying.